From 1ffff57e44a4a507786f5626070d9b1f3e371389 Mon Sep 17 00:00:00 2001
From: Juliusz Chroboczek <jch@irif.fr>
Date: Fri, 11 May 2018 14:13:06 +0200
Subject: [PATCH] Fix parsing of source length in filters.

This fixes a bug that was introduced in commit 4f4e3cb, and prevented
non-source-specific IPv4 routes from being redistributed.  Thanks
to Niklas Yann Wettengel for the detective work.
---
 configuration.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/configuration.c b/configuration.c
index 449e580..5f52ecb 100644
--- a/configuration.c
+++ b/configuration.c
@@ -479,10 +479,14 @@ parse_filter(int c, gnc_t gnc, void *closure, struct filter **filter_return)
            filter->src_plen_le < 128 || filter->src_plen_ge > 0)
             filter->af = AF_INET6;
     } else if(filter->af == AF_INET) {
-        filter->plen_le += 96;
-        filter->plen_ge += 96;
-        filter->src_plen_le += 96;
-        filter->src_plen_ge += 96;
+        if(filter->plen_le < 128)
+            filter->plen_le += 96;
+        if(filter->plen_ge > 0)
+            filter->plen_ge += 96;
+        if(filter->src_plen_le < 128)
+            filter->src_plen_le += 96;
+        if(filter->src_plen_ge > 0)
+            filter->src_plen_ge += 96;
     }
     *filter_return = filter;
     return c;
-- 
2.17.0

